invenio_app_ils/circulation/mail.py

```diff
@@ -0,0 +1,117 @@
+    """Templated message using Jinja2 blocks."""
+
+    def __init__(self, template_file, ctx={}, **kwargs):
+        """Build message body and HTML based on the provided template.
+
+        The template needs to provide two blocks: subject and body. An optional
+        html block can also be provided for HTML rendered emails.
+
+        :param template_file: Path to the template file.
+        :param ctx: A mapping containing additional information passed to the
+            template.
+        :param **kwargs: Keyword arguments as defined in
+            :class:`flask_mail.Message`.
+        """
+        self.template_file = template_file
+        template = current_app.jinja_env.get_template(template_file)
+        kwargs["subject"] = self.render_block(template, "subject", ctx)
+        kwargs["body"] = self.render_block(template, "body", ctx)
+        try:
+            kwargs["html"] = self.render_block(template, "html", ctx)
+        except TemplateError:
+            kwargs["html"] = kwargs["body"]
+        super(BlockTemplatedMessage, self).__init__(**kwargs)
+
+    @classmethod
+    def render_block(cls, template, block_name, ctx):
+        """Return a Jinja2 block as a string."""
+        new_context = template.new_context
+        if block_name not in template.blocks:
+            raise TemplateError("No block with name '{}'".format(block_name))
+        lines = template.blocks[block_name](new_context(vars=ctx))
+        return "\n".join(lines).strip()
+
+
+class LoanMessage(BlockTemplatedMessage):
+    """Email message for loan objects."""
+
+    def __init__(self, prev_loan, loan, **kwargs):
+        """Instantiate a new loan message.
+
+        :param prev_loan: Previous loan object.
+        :param loan: Updated loan object.
+        """
+        self.prev_loan = prev_loan
+        self.loan = loan
+        templates = current_app.config["LOAN_EMAIL_TEMPLATES"]
+        sender_email = current_app.config["MAIL_NOTIFY_SENDER"]
+        bcc = current_app.config["MAIL_NOTIFY_BCC"]
+        cc = current_app.config["MAIL_NOTIFY_CC"]
+        super(LoanMessage, self).__init__(
+            template_file=templates[self.get_template_name(prev_loan, loan)],
+            ctx=dict(
+                prev_loan=prev_loan,
+                loan=loan
+            ),
+            sender=kwargs.pop("sender", sender_email),
+            bcc=kwargs.pop("bcc", bcc),
+            cc=kwargs.pop("cc", cc),
+            **kwargs
+        )
+
+    @classmethod
+    def get_template_name(cls, prev_loan, loan):
+        """Get the template name for the loan transition."""
+        prev_state, state = prev_loan["state"], loan["state"]
+        name = state
+        if state == "PENDING":
+            available_item = get_available_item_by_doc_pid(loan["document_pid"])
+            if not available_item:
+                name = "LOAN_REQUESTED"
+        elif prev_state == state == "ITEM_ON_LOAN":
+            name = "EXTENDED"
```

Comment:
Do we manage somehow the case when the extension is blocked because the count exceeded max extensions allowed?